Add noteblocks and instruments

Paul1365972 opened the pull request:
This PR adds the noteblock and all blocks necessary to at play each instrument.
Not all blocks are implemented like logs or planks, but substibutes exist for each instrument (see: https://minecraft.wiki/w/Note_Block#Instruments).

This is still WIP for now and does not work with the redpiler yet.

Also small warning: The World trait now has a play_sound function, this will probably be a breaking change some software I don't know about. However adding it to the World trait still seemed like the best choice architectually.

BramOtte commented:
If you are concerned changing the World trait will break some other application you should add a default implementation to the new function you are adding.

StackDoubleFlow commented on crates/blocks/src/blocks/props.rs:
```diff
@@ -390,3 +390,159 @@ impl FromStr for TrapdoorHalf {
+            Instrument::Didgeridoo => "didgeridoo".to_owned(),
+            Instrument::Bit => "bit".to_owned(),
+            Instrument::Banjo => "banjo".to_owned(),
+            Instrument::Pling => "pling".to_owned(),
```
Here it's probably better to factor out the `to_owned` outside of the match.

StackDoubleFlow reviewed (commented):
Out of curiosity, do you know if this would negatively impact performance after #129?

Paul1365972 commented:
Rebased and benchmarked against current master. Performance is suprisingly enough exacltly the same.

StackDoubleFlow commented on crates/blocks/src/blocks/props.rs:
```diff
@@ -390,3 +390,160 @@ impl FromStr for TrapdoorHalf {
+    }
+}
+
+pub fn noteblock_note_to_pitch(note: u32) -> f32 {
```
Would it be better to keep this in `core::redstone::noteblock`? It seems more related to functionality of noteblocks rather than its properties of the block itself.

StackDoubleFlow commented on crates/core/src/redstone/mod.rs:
```diff
@@ -217,6 +218,36 @@ pub fn update(block: Block, world: &mut impl World, pos: BlockPos) {
+                    powered: should_be_powered,
+                };
+
+                // TODO: There is a weird double activation bug, so we need to check if block is not already powered
```
What is going on here? I'm not sure what you mean by double activation bug. Is it simply that the block updates multiple times? That's no guarantee it'll update only once after any change in surrounding state, or in response to a state change in general.

We already have `block` and `powered`. That state shouldn't have changed by this point, so I don't see why we need to retrieve and check it again.

Paul1365972 replied:
The block that is passed to the update method can definitely be outdated.
You are right that in nearly all cases the block is up-to-date as it fetched right before the update call, however in the `update` call in `turbo::breadth_first_walk` an old block instance is sometimes used.

Paul1365972 replied:
I looked into it a bit more and the comment on line 303 of `turbo.rs` basically explains it.

> This only works because updating any other block than a wire will never change the state of the block. If that changes in the future, the cached state will need to be updated

Noteblocks break that assumption as they do change state on update.
All other blocks are correctly unaffected by "double updates", either because they schedule a tick (repeaters, torches etc.) or because only the final state matters (trapdoors etc.).

StackDoubleFlow commented on crates/core/src/redpiler/backend/direct.rs:
```diff
@@ -679,7 +721,12 @@ fn get_all_input(node: &Node) -> (u8, u8) {
-fn update_node(scheduler: &mut TickScheduler, nodes: &mut Nodes, node_id: NodeId) {
+fn update_node(
+    scheduler: &mut TickScheduler,
+    sound_events: &mut Vec<Event>,
```
nit: `Event` is more general than just a sound event, so param should just be `events`

Paul1365972 commented:
Fixed all conflicts with the refactored redpiler and also updated the logic where necessary.
Feel free to re-review when you find the time.
I also changed the direct backend a bit so now each noteblock gets assigned a unique u16 id that is used to lookup the BlockPos, instrument and pitch (16 chunks full of noteblocks should be more than enough, I hope). For reference, originally it used a hashmap and the noteid as key.
Not sure if you want to have a look at the turbo backend issue, but my small hack still seems to work fine for now.

StackDoubleFlow reviewed (approved):
LGTM

The turbo wire impl should probably change wrt its incorrect assumption, but I don't think it's that important right now.
